packages/firestore/src/api/credentials.ts

```diff
@@ -188,44 +210,57 @@ export class FirebaseCredentialsProvider implements CredentialsProvider {
   private tokenCounter = 0;
 
   /** The listener registered with setChangeListener(). */
-  private changeListener: CredentialChangeListener | null = null;
+  private changeListener: CredentialChangeListener = () => Promise.resolve();
+
+  private invokeChangeListener = false;
 
   private forceRefresh = false;
 
-  private auth: FirebaseAuthInternal | null;
+  private auth: FirebaseAuthInternal | null = null;
+
+  private asyncQueue: AsyncQueue | null = null;
 
   constructor(authProvider: Provider<FirebaseAuthInternalName>) {
     this.tokenListener = () => {
       this.tokenCounter++;
+      this.receivedUser.resolve();
       this.currentUser = this.getUser();
-      this.receivedInitialUser = true;
-      if (this.changeListener) {
-        this.changeListener(this.currentUser);
+      if (this.invokeChangeListener) {
+        this.asyncQueue!.enqueueRetryable(() =>
+          this.changeListener(this.currentUser)
+        );
       }
     };
 
     this.tokenCounter = 0;
 
-    this.auth = authProvider.getImmediate({ optional: true });
+    const registerAuth = (auth: FirebaseAuthInternal): void => {
+      logDebug('FirebaseCredentialsProvider', 'Auth detected');
+      this.auth = auth;
+      this.awaitTokenAndRaiseInitialEvent();
+      this.auth.addAuthTokenListener(this.tokenListener);
+    };
 
-    if (this.auth) {
-      this.auth.addAuthTokenListener(this.tokenListener!);
-    } else {
-      // if auth is not available, invoke tokenListener once with null token
-      this.tokenListener(null);
-      authProvider.get().then(
-        auth => {
-          this.auth = auth;
-          if (this.tokenListener) {
-            // tokenListener can be removed by removeChangeListener()
-            this.auth.addAuthTokenListener(this.tokenListener);
-          }
-        },
-        () => {
-          /* this.authProvider.get() never rejects */
+    authProvider.onInit(auth => registerAuth(auth));
+
+    // Our users can initialize Auth right after Firestore, so we give it
+    // a chance to register itself with the component framework before we
+    // determine whether to start up in unauthenticated mode.
+    setTimeout(() => {
+      if (!this.auth) {
+        const auth = authProvider.getImmediate({ optional: true });
+        if (auth) {
+          registerAuth(auth);
+        } else if (this.tokenListener) {
+          // If auth is still not available, invoke tokenListener once with null
+          // token
+          logDebug('FirebaseCredentialsProvider', 'Auth not yet detected');
+          this.tokenListener();
         }
-      );
-    }
+      }
+    }, 0);
+
+    this.awaitTokenAndRaiseInitialEvent();
```

Comment:
Is `this.asyncQueue` set at this point? 

My understanding is it is not, so `awaitTokenAndRaiseInitialEvent` doesn't do anything except for setting `this.invokeChangeListener = false;` which will lead to wrong behavior if auth is not available - the following code in  this.tokenListener` will not execute:
```
      if (this.invokeChangeListener) {
        this.asyncQueue!.enqueueRetryable(() =>
          this.changeListener(this.currentUser)
        );
      }
```

Reply:
This seems to work in code though. Firestore will not receive a null user when auth is not initialized, and nothing seems to break.

Reply:
I changed the logic for `invokeChangeListener` to only set it to true if the AsyncQueue is available. This should address this concern.